classes/core/PKPPageRouter.php

```diff
@@ -522,14 +521,33 @@ private function _setLocale(PKPRequest $request, ?string $setLocale): void
             $request->setCookieVar('currentLocale', $newLocale);
         }
 
-        // Do not permit basic auth strings (user:password@url) in source parameter for redirects
-        if (preg_match('#^/\w#', $source = str_replace('@', '', $request->getUserVar('source') ?? ''))) {
-            $request->redirectUrl($source);
+        $indexUrl = $this->getIndexUrl($request);
+        if (isset($setLocale)) {
+            $targetUrl = '';
+            // The source parameter is coming either from the languageToggle block plugin
+            // and contains $smarty.server.SERVER_NAME|cat:$smarty.server.REQUEST_URI
+            // i.e. the whole URL except the protocol,
+            // or from TopNavActions and contain the whole URL with protocol
+            $source = $request->getUserVar('source');
+            if (isset($source)) {
+                $targetUrl = str_replace('@', '', $source);
+                $protocol = $request->getProtocol();
+                if (!str_starts_with($source, $protocol)) {
```

Comment:
This could, I think, be used to redirect to an arbitrary domain if the leading subdomain looks like a protocol, e.g. `http.myevilsite.com`.

Reply:
It will not be directly redirected to this source or targetUrl -- later in the code our index URL is assumed to be part of it, but I will maybe however add a check here for allowed hosts as well...